README
The ti (Time Index) and tis (Time Indexed Series) classes in the 
package provide date arithmetic facilities and an alternative to the somewhat
inflexible ts class in the standard R stats package. 

Time Indexes (ti class)

A time index has two parts: a tif (Time Index Frequency) code and a period.  tif
codes all lie in the open interval (1000..5000) and the period is a
nonnegative number less than 1e10.  The ti encodes both, as for any ti z

unclass(z) == (tif(z) * 1e10) + period(z)

Each tif has a particular base period (the time period where period(z) == 0).
For example, the base period for an anndecember (annual December) ti is the
year ending December 31, 1599.  Periods before the base period cannot be
represented by instances of the ti class.

If x and y are ti objects with the same tif, then 

x - y == period(x) - period(y)

and

x + (period(y) - period(x)) == y 

are both TRUE, so you can use ti's for various calendrical calculations. 

A jul class is also provided for Julian date-time objects. The jul()
constructor can create a jul from an ssDate (spreadsheet date), a
POSIXct or POSIXlt object, a ti object, a decimal time (like 2007.5), a
yyyymmdd number, a Date, or anything that can be coerced to a Date by
as.Date.  The ymd() function and its derivatives (year(), month(), day(), etc.)
work on anything that jul() can handle.


Time Indexed Series (tis class)

The tis class maps very closely to the FAME (http://www.sungard.com/Fame/) 
database notion of what a time series is.  A tis (Time Indexed Series) is
vector or matrix indexed by a ti.  If x is a tis, then start(x) gives the ti
for the first observation, and [start(x) + k] is the ti for the k'th
observation, while end(x) gives the ti for the last observation.  

You can replace, say, the 5'th observation in a tis x by

xstart <- start(x)
x[xstart + 4] <- 42

and of course the [ operator also works with a ti.  So if you want the value
of the daily series x from July 3, 1998, you can get it with

x[ti(19980703, "daily")]

provided, of course, that ymd(start(x)) <= 19980703 <= ymd(end(x)).

Numerous methods for tis objects are provided:
> methods(class = "tis")
 [1] aggregate.tis*     as.data.frame.tis* as.matrix.tis*     as.tis.tis*       
 [5] as.ts.tis*         cbind.tis*         cummax.tis*        cummin.tis*       
 [9] cumprod.tis*       cumsum.tis*        cycle.tis*         deltat.tis*       
[13] diff.tis*          edit.tis*          end.tis*           frequency.tis*    
[17] lag.tis*           lines.tis*         Ops.tis*           points.tis*       
[21] print.tis*         RowMeans.tis*      RowSums.tis*       start.tis*        
[25] tifName.tis*       tif.tis*           time.tis*          [<-.tis*          
[29] [.tis*             ti.tis*            t.tis              window.tis*       

as well as tisMerge() and tisPlot().  The convert() function change create
series of different frequencies from a given series in ways both more powerful
and more flexible than aggregate() can.



Setting Default Frequencies: 

Like the FAME DBMS, the ti class has seven weekly frequencies for weeks ending on
Sunday, Monday, and it has 12 annual frequencies, for years ending on the last
day of each of the 12 months. There are also multiple biweekly, bimonthly and
semiannual frequencies.  
At any time you can use the function setDefaultFrequencies() to change which
actual frequencies the strings "weekly", "biweekly", "bimonthly",
"quarterly", "semiannual" and "annual" refer to.  Note (as shown by
args(setDefaultFrequencies)) that if you don't specify some other frequencies
to setDefaultFrequencies(), you'll get default weeks ending on Monday and
default biweeks ending on the second Wednesday.  I chose these because they
are the weeks and biweeks used most often (for money and reserves data) in my
section at the Federal Reserve Board.  You might want to use something like

setHook(packageEvent("tis", "onLoad"), tis::setDefaultFrequencies(yourArgsGoHere))

in your .First() to automatically set the defaults up the way you want them
whenever this package is loaded.


LAGS:

The stats::lag(x, k) function says that a series lagged by a positive k starts
earlier.  The opposite is true for the Lag function in this package, to
maintain consistency with the common usage of 'first lag, second lag' and so
on in econometrics.

tisFilter:

The stats::filter() function coerces it's argument to a "ts" time series and
returns a "ts".  For "tis" and "zoo" series, this is not right.  Both I and
the author of the zoo package have requested that stats::filter be made an S3
generic with the current version renamed as filter.default.  This would allow
zoo and tis to define filter.zoo and filter.tis such that filter(aZooOrTis)
would do the right thing.  We are hopeful that this will happen soon.
Meanwhile, tisFilter() should be used to filter() a tis.



NEWS
Changes in Version 1.34

  o Add tis_init.c to eliminate note in cran checks

Changes in Version 1.33

  o adjusted convert to properly account for period bounderies when using businessday   
    data types from fame

  o Removed format.POSIXct, format.POSIXlt, format.default from being exported

Changes in Version 1.32

  o tisPlot() and scatterPlot() parameters xExpandBy, leftExpandBy,
    rightExpandBy and yExpandBy can now be a vector of two numbers to expand the
    data range asymmetrically.

  o tisFromCsv() was bombing when dateFormat = "excel" was given.

  o created a non-exported is.numeric.ti() that returns FALSE and explicitly
    exported as.Date.jul and as.Date.ti to prevent the zoo package's 
    reimplementation of the generic as.Date from incorrectly sending ti
    arguments to zoo::as.Date.numeric(). 

Changes in Version 1.31

  o Added 'offset' arg to as.Date.ti, and 'origin' arg to as.Date.jul. Request
    of Andre Mikulec.

  o Fixed minor bug in barplot.tis()

Changes in Version 1.30

  o Added tis methods for head() and tail().

Changes in Version 1.29

  o Removed aggregate.ts() function, which was a modified version of the same-named function
    in package:stats. Also modified aggregate.tis() to prefer a local version of
    aggregate.ts to the one in package:stats.

  o Registered S3 method t.tis()

Changes in Version 1.28

  o Added tz argument to tisFromCsv() and added logic to handle date strings that can't be
    turned into valid dates. Thanks again to Ying Leung for reporting the problem that led
    to the discovery of a bug.

  o Added naLoses argument to mergeSeries(), with default value of FALSE for backward compatibility.

Changes in Version 1.27

  o Fixed bug in tisFromCsv() reported by Ying Leung. Also added chopNAs parameter to that
    function.

  o Changed name of ymd.xy() to ymdXy() to keep the CRAN checker happy.

Changes in Version 1.26

  o now exports the barplot2() function called by barplot.tis

Changes in Version 1.25

  o new capitalize() function moved here from the another package.

  o new functions ymd.xy() and ymdShade() generalize the old nber.xy() (now defunct) and
    nberShade() functions to shade arbitrary date ranges on tisPlots, not just NBER
    recessions.

  o changed names of "nber" arguments to tisPlot() to reflect the more general date range
    shading functionality.

Changes in Version 1.24

  o freq2tif() gave wrong answer for frequency = 261, business day freq.

  o tierChart(vlines.month) now draws vertical line after the last day of the month, rather
    than after the first day of the month.

Changes in Version 1.23

  o added some support for IDate objects from the data.table package.

  o added isBusinessDay() function at the suggestion of Adam Hogan.

  o fixed a bug in inferDate() that was causing tisFromCsv() to misbehave

Changes in Version 1.22

  o defined some variables locally in nberShade.ggplot() to avoid spurious R
    CMD check warnings.

Changes in Version 1.21

  o changed name of as.list.keepClass() to asClassyList() to placate CRAN

Changes in Version 1.20

  o fixed bug in assignList()

Changes in Version 1.19

  o added xOffset argument to tisPlot()

  o fixed a bug in nberShade.ggplot() found by Matthew McCormick. Thanks, Matt.

Changes in Version 1.18

  o added a tis method for the generic barplot() function

  o moved tierChart functions from local frb package to tis package

Changes in Version 1.17

  o ymdToTi() (an internal function in ti.R) for frequencies bmnovember,
    qoctober and qnovember was giving results that were off by one period,
    causing convert() to give wrong results when aggregating to those
    frequencies. Thanks again to Abiel Reinhart for reporting the problem.

  o changed "env = envir" to "envir = envir" in several places to avoid
    package check complaints about partial matches

Changes in Version 1.16

  o more bugs in convert() reported by Abiel Reinhart

Changes in Version 1.15

  o fixed bug in convert() reported by Abiel Reinhart

Changes in Version 1.14

  o added as.POSIXct.jul and as.POSIXct.ti methods

Changes in Version 1.13

  o make "[<-.tis" a bit smarter when only a single index is supplied

Changes in Version 1.12

  o fixed bugs in "[.tis" and "[<-.tis" reported by Abiel Reinhart.

Changes in Version 1.11

  o added 'zero' argument to naWindow

Changes in Version 1.10

  o added methods for generic function xtfrm for classes jul, ssDate, ti and
    tis.

  o print.ti() and print.tis() can now suppress the 'class:  ti' and 'class:
    tis' strings, respectively, at the end.

Changes in Version 1.9

  o convert() uses tapply() which calls as.factor() on a ti object. For some
    reason, this results in NA's in R 2.10.0.  Unclassing the ti before using
    it as a parameter to tapply fixed it.  Why factor() gives NA's for ti
    objects is a subject for further investigation.

Changes in Version 1.8

  o Allow fortify.tis to handle proper naming when a univariate array is 
    converted into a tis object.  Contributed by Trevor Davis.

Changes in Version 1.7

  o Tweaked nberShade.ggplot to work with a greater range of ggplot objects,
    and improved examples for nberShade and fortify.tis documentation.
    Contributed by Trevor Davis.

Changes in Version 1.6 (contributed by Trevor Davis)

  o made nberShade() generic and added a method for ggplot objects

  o added fortify.tis(), a tis-specific method for the fortify() generic in ggplot2

Changes in Version 1.5

  o added options to how nberDates() and nber.xy() handle an ongoing recession

Changes in Version 1.4

  o fixed nberDates() to show latest recession

Changes in Version 1.3

  o cleaned up some documentation to incorrect package references and
    references to Fed-specific stuff.

  o added "tif" parameter to today() with default value "daily"

Changes in Version 1.2

  o better handling of ti objects with some NA values.  Thanks to Chuck

R/tier.R
tierChart <- function(x,
                      startMonth = latestJanuary(end(x)),
                      nMonths = 4,
                      nYears = 7,
                      offsets = 0,
                      padDays = 6,
                      pch = "year", 
                      lty = "solid",
                      lwd = 1.5, 
                      col = 1 + (n.y:1),
                      type = "b", 
                      ylim = NULL,
                      outlier.trim = 0,
                      noTrimLastYear = TRUE,
                      extendHorizontalTicks = TRUE,
                      circles.ymd = NULL,
                      circles.col = 6,
                      circles.inches = 0.1,
                      vlines.ymd = NULL,
                      vlines.col = 2,
                      vlines.lty = 4,
                      vlines.lwd = 1.5,
                      vlines2.ymd = NULL,
                      vlines2.col = 3,
                      vlines2.lty = "solid",
                      vlines2.lwd = 2,
                      hlines = NULL,
                      hlines.col = 1,
                      hlines.lty = 1,
                      hlines.lwd = 1,
                      tiPoints.1 = NULL,
                      tiPoints.2 = NULL,
                      pch.1 = "*", 
                      pch.2 = "+",
                      col.1 = 2, 
                      col.2 = 3,
                      nolegend = FALSE,
                      main = deparse(substitute(x)),
                      topleft.labels = NULL,
                      topright.labels = NULL,
                      legend.ncol = length(years),
                      legend.bg = 0,
                      timestamp = TRUE,
                      topline = TRUE,
                      vlines.periodEnd = TRUE,
                      vlines.month  = TRUE,
                      midperiod = FALSE,
                      lwdLastYear = 1.5,
                      cex = 1.5,
                      boxes = TRUE,
                      ...){
  ## vlines.monday has been replaced by vlines.periodEnd, but it remains a
  ## valid parameter name for the sake of backwards compatibility
  
  ## graphics parameters
  oldpar <-  par(xpd = FALSE, err = -1, mar = c(2,5,3,5)+0.1, mgp = c(1,0,0),
                 xaxs = "i", yaxs = "i", mex = 1, ann = FALSE)
  on.exit(par(oldpar), add = TRUE) 
  if(par("new") == FALSE) plot.new()
  main <- main
  ## The data lines and points for tier charts are drawn by the matplot()
  ## function. We start by creating  matrices with columns for each
  ## year plotted. We'll need:
  ##   py -- y coordinates of the points to be plotted
  ##   px -- x coordinates of the points to be plotted
  ##   pymd -- ymd dates corresponding to the plotted points
  ##   pindex -- subscripts for x, i.e., px[,i] == x[pindex[,i]]  
  ##             (except for NA values in pindex)
  ## Then we'll draw the plot and return.
  
  x  <- as.tis(x)
  xTif <- tif(x)
  dim(x) <- NULL
  if(!is.ti(startMonth)) stop("startMonth must be a ti")
  xEndMonth <- currentMonth(end(x))
  if(startMonth > xEndMonth) stop("startMonth after end(x)")
  endMonth <- startMonth + nMonths - 1
  if(endMonth > xEndMonth){
    xExtendedEnd <- ti(endMonth, tif(x))
    x[xExtendedEnd] <- NA
  }
  xStartMonth <- currentMonth(start(x))
  
  years  <- (year(startMonth) - nYears + 1): year(startMonth)
  eyears <- (year(endMonth)   - nYears + 1): year(endMonth)
  
  eym <- 100*eyears + month(endMonth)
  years <- years[eym >= (ymd(start(x)) %/% 100)]

  ## julian dates of the series
  if(midperiod) xj <- floor(jul(ti(x), offset = 0.5))
  else          xj <- jul(ti(x))

  ## some lengths and sequence vectors
  n.y      <- length(years)
  n.x      <- length(x)
  y.seq    <- 1:n.y
  x.seq    <- 1:n.x

  startMonths <- startMonth - 12*((nYears - 1):0)
  endMonths   <- startMonths + nMonths - 1
  ## drop (start,end) pairs outside the range of data
  inRange <- startMonths <= xEndMonth & endMonths >= xStartMonth
  startMonths <- startMonths[inRange]
  endMonths   <- endMonths[inRange]
  ## starts and ends: vectors of starting and ending Julian dates for each year
  ## on the plot
  offsets <- c(rev(offsets), rep(0, length(years) - length(offsets)))
  starts <- jul(startMonths - 1) - (padDays - 1) + offsets
  ends   <- jul(endMonths) + padDays + offsets

  ## the x-axis starts at 1 and has daily ticks
  x.ticks <- 1:(1 + ends[n.y] - starts[n.y])
  ## year[n.y] ymd dates associated with x.ticks
  x.ymd <- ymd(seq(starts[n.y], ends[n.y]))
  ## x.periodEnds[i] is TRUE iff x.ymd[i] is the last day of a period of x
  x.periodEnds <- x.ymd == ymd(ti(x.ymd, tif = xTif))
  
  ## If Feb 28 appears on the plot, Feb 29 has to as well. If it's not already
  ## there, add a fake observation to x.ticks, x.ymd and x.periodEnds to
  ## accomodate it. 
  feb28 <- match(0228, x.ymd %% 10000, nomatch = FALSE)
  if(feb28 != FALSE){
    feb29 <- match(0229, x.ymd %% 10000, nomatch = FALSE)
    if(feb29 == FALSE){
      x.ticks       <- append(x.ticks, length(x.ticks) + 1)
      x.ymd         <- append(x.ymd, x.ymd[feb28]+1, after = feb28)
      x.periodEnds  <- append(x.periodEnds, FALSE, after = feb28)
    }
  }
  xtlen <- length(x.ticks)
  
  ## initialize matrices with NA's 
  matrows <- ceiling(xtlen*frequency(x)/365) + 1
  py <- matrix(NA, nrow = matrows, ncol =  n.y,
               dimnames = list(character(0), paste(years)))  
  pindex <- pymd  <- px <- py
  
  ## fill in matrices by looping thru the years
  for(y in y.seq){ 
    ix <- x.seq[between(xj, starts[y], ends[y])]
    iRows <- 1:length(ix)  ## Rows to fill in

    pindex[iRows, y] <- ix
        py[iRows, y] <- x[ix]
      pymd[iRows, y] <- ymd(xj[ix])
          ymd.offset <- 10000 * (max(years) - years[y])
        px[iRows, y] <- match(ymd(xj[ix] + offsets[y]) + ymd.offset, x.ymd)
  }
  ## The plot is drawn in six steps. The first is to draw a blank plot, solely
  ## for the purpose of setting the plot coordinates.  Next, we add the legends
  ## and titles, and only then do we actually draw the data lines and
  ## symbols. Reversing the usual order this way allows the data lines to
  ## overwrite the legends and titles. The fourth step is to draw the grid lines
  ## and the axes. Fifth is adding in the optional "extra" points and vertical
  ## lines specified by tiPoints.1, tiPoints.2, and vlines.ymd. The sixth and
  ## final step is to draw circles at the points indicated by circles.ymd

  ## Step 1: Setting coordinates
  xlim <- c(x.ticks[1], x.ticks[xtlen] + 1)
  ##   The y coordinates are governed by two parameters: outlier.trim and ylim.
  ##   If ylims are set explicitly, outlier.trim has no effect.  Otherwise, ylim
  ##   will be calculated by making sure that the 
  ##   c(outlier.trim, 1 - outlier.trim) quantiles fall within ylim.  The
  ##   default value (outlier.trim = 0) makes sure all observations fall within
  ##   the plot area.
  ##   In either case, ylim will be further stretched (if necessary) to
  ##   accomodate "pretty" y-axis labels.
  if(missing(ylim)){
    bounds <- 1.1 * quantile(py, c(outlier.trim, 1-outlier.trim), na.rm=TRUE) - 
              0.1 * median(py, na.rm=TRUE)
    ylim <- c(max(bounds[1], min(py, na.rm=TRUE)), 
              min(bounds[2], max(py, na.rm=TRUE)))
    if(noTrimLastYear){
      pyLastYear <- py[,ncol(py)]
      ylim <- c(min(bounds[1], min(pyLastYear, na.rm = TRUE)),
                max(bounds[2], max(pyLastYear, na.rm = TRUE)))
    }
  }
  ylim[2] <- 1.1*ylim[2] - 0.1*ylim[1]
  y.ticks <- pretty(ylim, n = 8)
  ylim <- range(y.ticks)
  ##   draw the blank plot.
  plot.window(xlim = xlim, ylim = ylim)

  
  ## Step 2: Add legends and titles
  if( length(pch) == 1 && is.character(pch) && pch == "year")
    pch <- paste(substring(years, 4, 4), collapse = "")

  if(!missing(lwdLastYear)){
    lwd <- rep(lwd, length = length(years))
    lwd[length(lwd)] <- lwdLastYear
  }
  
  if(!nolegend){
    legendStrings <- format(years)
    if(any(offsets != 0)){
      roffStrings <- paste("[", offsets, "]", sep = "")
      roffStrings[offsets == 0] <- ""
      legendStrings <- paste(legendStrings, roffStrings, sep = "")
    }
    legend(x = 0.5*xlim[1] + 0.5*xlim[2],
           y = max(ylim),
           legend = legendStrings, 
           pch = pch,
           lty = lty,
           lwd = lwd,
           col = col,
           ncol = legend.ncol,
           bg = legend.bg, bty = "o", cex = 1.3,
           xjust = 0.5)  
  }

  if(length(main)) 
	mtext(main, side = 3, adj = 0.5, cex = 1.5, line = 0.5)	
  ##   upper left labels
  if(tl.len <- length(topleft.labels)){
    for(i in 1:tl.len)
      mtext(topleft.labels[i], side = 3, adj = 0, at = xlim[1], 
	    line = tl.len - i + 0.2)
  }
  ##   upper right labels
  if(tr.len <- length(topright.labels)){
    for(i in 1:tr.len)
      mtext(topright.labels[i], side = 3, adj = 1, at = xlim[2], 
	    line = tr.len - i + 0.2)
  }
  
  ## Step 3: Data lines and points
  par(new = TRUE) ## don't erase what we've already drawn
  matplot(px, py, type = type, pch = pch, col = col, cex = cex, 
          lty = lty, lwd = lwd, axes = FALSE, bty = "n", yaxs = "i", xaxs = "i", 
          xlim = xlim, ylim = range(y.ticks), ...) 
  if(type == "b" | type == "p") 
    matpoints(px, py, pch = pch, lty = 1, col = col, cex = cex)
  
  ## Step 4: Grid lines and axes
  ##   locations of period ends, and days 1 and 15 of a month
  periodEnd.ticks  <- x.ticks[x.periodEnds]
  day15 <- (1:xtlen)[x.ymd %% 100 == 15]
  eom   <- (1:xtlen)[x.ymd == ymd(ti(x.ymd, tif = "monthly"))]
  ##   vertical periodEnd dashed lines
  if(vlines.periodEnd)
    axis(1, at = periodEnd.ticks, tck = 1, lty = 2, labels = FALSE)
  ##   vertical last day of month grid lines
  if(vlines.month) abline(v = eom, lty = 1)
  ##   left axis
  if(extendHorizontalTicks)
    axis(2, at = y.ticks, tck = 1, lty = 2, labels = FALSE) 
  axis(2, at = y.ticks, las = 1, adj = 1, tck = 0.02, labels = TRUE) 
  ##   right axis
  axis(4, at = y.ticks,  las = 1, adj = 0, tck = 0.02) 
  ##   top axis
  if(topline) abline(h = ylim[2])
  ##   bottom axis
  axis(1, at = x.ticks, tck = 0.02, labels = FALSE)

  ##   if the plot covers 8 or more months, use abbreviated month labels
  ##   only. Otherwise use full month labels and PeriodEnd labels
  monthLabels <- format(jul(x.ymd[day15]), "%b %Y")
  if(length(monthLabels) > 8)
    axis(1, at = periodEnd.ticks,  tck = 0.04, labels = FALSE)
  else {
    periodEndLabels <- paste(x.ymd[x.periodEnds] %% 100)
    axis(1, at = periodEnd.ticks,  tck = 0.04, labels = periodEndLabels)
  }
  mtext(text = monthLabels, side = 1, line = 1, at = day15) 

  ## Step 5: "Extra" points and vlines
  if(!is.null(vlines.ymd)){  
    ## draw vertical lines at the dates in vlines.ymd
    i <- match(vlines.ymd, x.ymd, nomatch = 0)
    if(any(i != 0))
      abline(v  =  x.ticks[i], col = vlines.col, lty = vlines.lty, lwd = vlines.lwd)
  }
  if(!is.null(vlines2.ymd)){  
    ## draw vertical lines at the dates in vlines.ymd
    i <- match(vlines2.ymd, x.ymd, nomatch = 0)
    if(any(i != 0))
      abline(v  =  x.ticks[i], col = vlines2.col, lty = vlines2.lty, lwd = vlines2.lwd)
  }
  if(!is.null(hlines)){
    abline(h = hlines, col = hlines.col, lty = hlines.lty, lwd = hlines.lwd)
  }
  if(!is.null(tiPoints.1)){  
    ## add points to plot
    ymd.1 <- ymd(ti(tiPoints.1))
    i.1 <- match(ymd.1, x.ymd, nomatch = 0)
    if(any(i.1 != 0))
      matpoints(x.ticks[i.1], tiPoints.1[i.1 > 0], pch = pch.1,
		cex = 1.4, col = col.1)
  }

  if(!is.null(tiPoints.2)){  
    ## add even more points to plot
    ymd.2 <- ymd(ti(tiPoints.2))
    i.2 <- match(ymd.2, x.ymd, nomatch = 0)
    if(any(i.2 != 0))
      matpoints(x.ticks[i.2], tiPoints.2[i.2 > 0], pch = pch.2,
		cex = 1.4, col = col.2)
  }
  
  ## Step 6 : Circle the points whose dates match circles.ymd
  if(!is.null(circles.ymd)){
    i.circles <- match(circles.ymd, pymd, nomatch = 0)
    if(any(i.circles != 0)){
      x <- px[i.circles]
      y <- py[i.circles]
      symbols(x, y, add = TRUE, circles = rep(1, length(x)),
	      inches = circles.inches, col = circles.col)
    }
  }

  if(boxes){
    ## add arrow boxes 0.5 x 0.25 inches
    extents <- c(diff(xlim), diff(ylim)) / (par("pin") * c(2, 4))
    boxBottom <- ylim[2] - extents[2]
    lBox <- c(xlim[1], boxBottom, xlim[1]+extents[1], ylim[2])
    rBox <- c(xlim[2]-extents[1], boxBottom, xlim[2], ylim[2])
    rect(lBox[1], lBox[2], lBox[3], lBox[4],  col = "white")
    rect(rBox[1], rBox[2], rBox[3], rBox[4],  col = "white")
    arrows(lBox[3], (lBox[2] +lBox[4])/2, lBox[1], (lBox[2]+lBox[4])/2,
           angle = 30,lwd = 4, col = "black")
    arrows(rBox[1], (rBox[2] +rBox[4])/2, rBox[3], (rBox[2]+rBox[4])/2,
           angle = 30,lwd = 4, col = "black")
    printBox <- c(lBox[3], lBox[2], lBox[3] + (lBox[3] - lBox[1]), lBox[4])
    rect(printBox[1], printBox[2], printBox[3], printBox[4], col = "darkblue")
    text(x = mean(printBox[c(1,3)]),
         y = mean(printBox[c(2,4)]),
         labels = "Print Me",
         adj = c(0.5, 0.5),
         cex = 1.25, col = "white")
  }
  else {
    lBox <- rBox <- printBox <- NULL
  }
  
  ## Date-Time Stamp 
  if(timestamp) mtext(date(), side = 3, outer = TRUE, line = -1, adj = 1)
  
  invisible(list(px = px, py = py, ymd = pymd, index = pindex,
                 lBox = lBox, rBox = rBox, printBox = printBox,
                 startMonth = startMonth, nMonths = nMonths,
                 nYears = nYears))
}

adjustableTierChart <- function(x, ..., edit = TRUE,
                                changes = numeric(0), verbose = FALSE){
  locateOne <- function(){
    options(error = NULL)
    point <- try(locator(1), silent = TRUE)
    if(class(point) == "try-error")
      point <- NULL
    options(error = quote(recover()))
    point
  }

  arglist <- list(..., edit = edit, verbose = verbose)
  if(is.null(arglist$main)) 
    arglist$main <- deparse(substitute(x))
  arglist$x <- x
  arglist$changes <- changes
  arglist$boxes <- TRUE
  drops <- match(c("edit", "changes", "verbose"), names(arglist), nomatch = 0)
  tcArgs <- arglist[-drops]
  tc <- do.call("tierChart", tcArgs)

  xStart <- start(x)
  xEnd   <- end(x)
  xStartMonth <- ti(xStart, tif = "monthly")
  xEndMonth   <- ti(xEnd,   tif = "monthly")
  
  ## adjust
  lastcol    <- dim(tc$px)[2]
  not.na     <- !(is.na(tc$px[,lastcol]) | is.na(tc$py[,lastcol]))
  px         <- tc$px[not.na, lastcol]
  py         <- tc$py[not.na, lastcol]
  ymd        <- tc$ymd[not.na, lastcol]
  index      <- tc$index[not.na, lastcol]
  xStep      <- mean(diff(px))
  lBox       <- tc$lBox
  rBox       <- tc$rBox
  pBox       <- tc$printBox
  startMonth <- tc$startMonth
  nMonths    <- tc$nMonths
  scrollIncrement <- max(nMonths - 2, 1)
  if(frequency(xStart) <= 12) scrollIncrement <- round(frequency(xStart))
  
  ## refresh arrows for changes that are still in range
  if(length(changes) > 0){
    for(i in 1:nrow(changes)){
      change <- changes[i,]
      j <- match(change["ymd"], ymd, nomatch = 0)
      if(j > 0) arrows(px[j], py[j], px[j], change["newY"], col = 2)
    } 
  }
  
  if(verbose)  cat("Awaiting mouse input...")
  nextpoint  <- locateOne()
    
  while(length(nextpoint) != 0){
    npX <- nextpoint$x
    npY <- nextpoint$y
    lBoxClicked <- rBoxClicked <- pBoxClicked <- FALSE
    ## see if user clicked in lBox, rBox or pBox
    if((npY >= lBox[2]) && (npY <= lBox[4])){
      lBoxClicked <- (npX >= lBox[1]) && (npX <= lBox[3])
      rBoxClicked <- (npX >= rBox[1]) && (npX <= rBox[3])
      pBoxClicked <- (npX >= pBox[1]) && (npX <= pBox[3])
    }
    if(rBoxClicked){
      inc <- min(scrollIncrement, xEndMonth - startMonth)
      arglist$startMonth <- startMonth + inc
      return(do.call("adjustableTierChart", arglist))
    }      
    if(lBoxClicked){
      inc <- min(scrollIncrement, startMonth - xStartMonth)
      arglist$startMonth <- startMonth - inc
      return(do.call("adjustableTierChart", arglist))
    }
    if(pBoxClicked){
      printer <- options()$colorPrinter
      if(is.null(printer)) cat("options()$colorPrinter not set\n")
      else {
        postscript(file = "",
                   command = paste("lp -d", printer, sep = ""),
                   pointsize = 9)
        setColors(frColors())
        modArgs <- tcArgs
        modArgs$legend.bg <- "grey90"
        do.call("tierChart", modArgs)
        dev.off()
      }
    }
    else {
      if(edit){
        ## find nearest point (by x value)
        xDistances <- abs(px - npX)
        if(min(xDistances) < xStep){ 
          i <- order(xDistances)[1]
          ## change the yvalue
          new.y <- npY
          ## draw an arrow from old point to new point,
          ##   color it the same color 
          ##   as the most recent year plotted
          changes <- rbind(changes, c(ymd = ymd[i], newY = npY))
          arglist$changes <- changes
          arrows(px[i], py[i], px[i], new.y, col = 2)
          if(verbose)
            cat("Move", ymd[i], "obs from", signif(py[i], 3),
                "to", signif(new.y, 3), "\nAwaiting mouse input... ")
        }
      }
    }
    ## repeat
    nextpoint <- locateOne()
  }

  if(length(changes) > 0)
    x[match(changes[,"ymd"], ymd(ti(x)))] <- changes[,"newY"]
  retval <- x
  attr(retval, "startMonth") <- arglist$startMonth
  if(verbose) cat("adjustableTierChart() exiting.\n")
  invisible(retval) 
}

